### DISCUSSION TITLE
Can't assign new value to "Value" property at ScrollBar control

### DISCUSSION THREAD

**Discussion by well-coder** [AvaloniaUI/Avalonia#16192] [Q&A] [answered]
Hello everyone! 

I'm trying to restore vertical ScrollBar position after update data in my DataGrid.
Firstly, I take a ScrollBar from my DataGrid like this:

```
  private ScrollBar FindVerticalScrollBar()
  {
      var dataGrid = this.FindControl<DataGrid>("MyDataGrid");
      var verticalScrollBar = dataGrid.GetVisualDescendants().OfType<ScrollBar>().Single(bar => bar.Orientation == Orientation.Vertical);
      return verticalScrollBar;
  }
```
Then, i observe for changes on ScrollBar's property "Value"  and store it last position like

```
private void HandleActivation(CompositeDisposable disposables)
{
    _scroll = FindVerticalScrollBar();

    _scroll.GetObservable(RangeBase.ValueProperty)
        .Where(value => value != default)
        .SkipWhile(_ => _isBusy)
        .Subscribe(SaveLastPosition)
        .DisposeWith(disposables);

    ViewModel.ViewModelUpdated += RestorePosition;
}

    private void SaveLastPosition(double newValue)
    {
        _lastPosition = newValue;
    }
```

After my viewModel updated (it happens every 20 seconds by `Observable.Timer`), i wanna assign last position to "Value" property and restore it previous state

   ```
 private void RestorePosition()
    {
        _isBusy = true;
        _scroll.Value = _lastPosition;
        _isBusy = false;
    } 
```

But when i'm tyring to set  `_scroll.Value = _lastPosition;` nothing happens. _scroll.Value remains 0.
Where did I go wrong?

**Comment by workgroupengineering** [comment]
Try with

```c#
private void RestorePosition()
 {
     _isBusy = true;
     _scroll.SetCurrentValue(RangeBase.ValueProperty,  _lastPosition);
     _isBusy = false;
 } 
```

**Reply by well-coder** [reply]
doesn't work =(

**Reply by workgroupengineering** [reply]
Check that RestorePosition is called in the context of UI Thread.

```c#
private void RestorePosition()
 {
     System.Diagnostics.Debug.Assert(Avalonia.Threading.Dispatcher.UIThread.CheckAccess());
     _isBusy = true;
     _scroll.SetCurrentValue(RangeBase.ValueProperty,  _lastPosition);
     _isBusy = false;
 } 
```

**Reply by well-coder** [reply]
true

**Reply by timunie** [reply, marked as answer]
that's not as easy as it sounds like, due to virtualization. Better if you know which item to scroll into view. Then you can simply call `myDataGrid.BringIntoView(item)`

**Reply by well-coder** [reply]
Yes, I guess it will be better. Thank You!

**Reply by well-coder** [reply]
Note: It works fine only if called from `Dispatcher.UIThread.Post(someAction, DispatcherPriority.Background)`
